lib/GADS.pm

```diff
@@ -3024,6 +3025,89 @@ prefix '/:layout_name' => sub {
 
     };
 
+    any ['get', 'post'] => '/historic_purge/' => require_login sub {
+        my $layout = var('layout') or pass;
+
+        forwardHome({ danger => "You do not have permission to manage deleted records"}, '')
+            unless $layout->user_can("purge");
+
+        my $user        = logged_in_user;
+        my $view        = current_view($user, $layout);
+
+        my $records = GADS::Records->new(
+            user                => $user,
+            layout              => $layout,
+            schema              => schema,
+            include_children    => 1,
+            view                => $view,
+        );
+
+        my $params = +{};
+        my @record_data;
+
+        if(param 'submit') {
+            my @columns = body_parameters->get_all('column_id');
+            $records->columns(\@columns);
+        } elsif(param 'purge') {
+            my @columns = body_parameters->get_all('column_id');
+            $records->columns(\@columns);
+        }
+
+        unless(param 'purge') {
+            while (my $record = $records->single) {
+                foreach my $column (@{$record->columns_render}) {
+                    next if $column->internal;
+
+                    my $clone = $record->clone;
+                    my $count;
+                    my $current_id = $record->current_id;
+                    $clone->find_chronology_id($current_id);
+                    foreach my $cron (@{$clone->chronology}) {
+                        $count = scalar(grep {$_->{name} eq $column->name}  @{$cron->{changed}}) + 1;
+                    }
+                    
+                    my @pmc= $record->presentation_map_columns(columns=> [$column]);
+                    push @record_data, {
+                        col_id => $column->id,
+                        name => $column->name,
+                        id => $record->record->{id},
+                        value => $pmc[0]->{data}->{value}, # THIS SHOULD BE THE VALUE OF THE GIVEN RECORD/COLUMN
+                        count => $count || 0,
+                    };
+                }
+            }
+        }
+
+        if(param 'submit') {
+            $params->{records} = \@record_data;
+            my @mapped_ids = map {$_->{col_id}} @record_data;
+            my @mapped_records = uniq map {$_->{id}} @record_data;
+            $params->{column_list} = \@mapped_ids;
+            $params->{record_list} = \@mapped_records;
```

Comment:
I haven't looked at this in detail, but I think there could be some improvements here, to prevent having one parameter built from another parameter. I think probably better to have separate variables that are populated as required. E.g. push the record IDs onto a separate array in the $record loop rather than the $column loop. Similar for the column IDs, I think this could be `$params->{column_list} = [map $_->id, $records->columns_render]` or similar.

Reply:
Looking at this a bit deeper, I am wondering if I could do this in the view, rather than here using the `@record_data` (as that is already passed to the view) rather than passing the same thing 3 different ways - I'll take a look at both methods and work out which is best.